addons/payment_alipay/controllers/main.py

```diff
@@ -14,48 +15,107 @@
+        """ Process the notification data sent by Alipay after redirection from checkout.
+
+        See https://global.alipay.com/docs/ac/web/sync.
+
+        :param dict data: The notification data
+        """
         _logger.info("handling redirection from Alipay with data:\n%s", pprint.pformat(data))
+
+        # Check the integrity of the notification
+        tx_sudo = request.env['payment.transaction'].sudo()._get_tx_from_feedback_data(
+            'alipay', data
+        )
+        self._verify_notification_signature(data, tx_sudo)
+
+        # Handle the notification data
         request.env['payment.transaction'].sudo()._handle_feedback_data('alipay', data)
         return request.redirect('/payment/status')
 
-    @http.route(_notify_url, type='http', auth='public', methods=['POST'], csrf=False)
-    def alipay_notify(self, **post):
-        """ Alipay Notify """
-        _logger.info("notification received from Alipay with data:\n%s", pprint.pformat(post))
-        self._alipay_validate_notification(**post)
-        request.env['payment.transaction'].sudo()._handle_feedback_data('alipay', post)
-        return 'success'  # Return 'success' to stop receiving notifications for this tx
+    @http.route(_webhook_url, type='http', auth='public', methods=['POST'], csrf=False)
+    def alipay_webhook(self, **data):
+        """ Process the notification data sent by Alipay to the webhook.
 
-    def _alipay_validate_notification(self, **post):
-        tx_sudo = request.env['payment.transaction'].sudo()._get_tx_from_feedback_data(
-            'alipay', post
-        )
-        if not tx_sudo:
-            raise ValidationError(
-                "Alipay: " + _(
-                    "Received notification data with unknown reference:\n%s", pprint.pformat(post)
-                )
+        See https://global.alipay.com/docs/ac/web/async.
+
+        :param dict data: The notification data
+        :return: The 'SUCCESS' string to acknowledge the notification
+        :rtype: str
+        """
+        _logger.info("notification received from Alipay with data:\n%s", pprint.pformat(data))
+        try:
+            # Check the origin and integrity of the notification
+            tx_sudo = request.env['payment.transaction'].sudo()._get_tx_from_feedback_data(
+                'alipay', data
             )
+            self._verify_notification_origin(data, tx_sudo)
+            self._verify_notification_signature(data, tx_sudo)
 
-        # Ensure that the notification was sent by Alipay
-        # See https://global.alipay.com/docs/ac/wap/async
-        acquirer_sudo = tx_sudo.acquirer_id
-        val = {
+            # Handle the notification data
+            request.env['payment.transaction'].sudo()._handle_feedback_data('alipay', data)
+        except ValidationError:  # Acknowledge the notification to avoid getting spammed
+            _logger.exception("unable to handle the notification data; skipping to acknowledge")
```

Comment:
so we still return a `SUCCESS` in case we failed to handle it?

Reply:
Yes. Sometimes we'll receive notifications that will cause a ValidationError to be raised for a specific transaction. Since we don't want the provider to shut down the webhook, the async notification is discarded. We don't do that with synchronous notifications, though.